contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,208 @@ The central manager will poll each worker for their discovered volumes and creat
 them.  When a PV is released, then it will send the cleanup request to the worker.
 
 Detailed design TBD
+
+
+### Phase 3: 1.11 Alpha
+
+#### Dynamic Provisioning
+
+The local volume provisioner will be enhanced to support dynamic provisioning of local storage,
+which depends on the newly introduced changes in [Volume Topology-aware Scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+that is:
+
+1. When the feature gate `DynamicProvisioningScheduling` is enabled,
+the scheduler will verify a pod's volume requirements,
+and set an annotation `annSelectedNode` on PVCs that need provisioning.
+2. New fields `Capacity` will be added in `Status` of `StorageClass` to
+allow storage providers to report capacity limits per topology domain.
+
+The local volume provisioner needs to be updated in several areas to support dynamic provisioning.
+
+##### Backend Storage Management
+
+Mechanisms to manage local storage sources are varied.
+In addition to current `populator`, `deletor` and `discoverer`, a new component named
+`storageManager` will be introduced in the plugin, to manage the backend storage.
+
+According to the input of classes in the configuration, there will be a list of `wrappers`
+under the manager, to communicate with backend storage. `wrapper` is a general interface,
+each kind of storage should have its own implementation. storageManager store the wrappers
+according to storage classes, wrappers of different classes can be different.
+LVM wrapper will be introduced in alpha phase.
+
+```
+Wrappers map[string]wrapper.LocalVolumeWrapper
+```
+
+During start-up, a new environment variable is needed for the provisioner DaemonSet:
+
+* `PROVISIONER_NAME` to specify the name of the provisioner, its value matches the Provisioner
+specified in the StorageClass. It will be default to `local-volume-provisioner` if not specified.
+
+**Note:** Currently we already have a `provisionerName` in `RuntimeConfig` of the plugin,
+which is in format of `local-volume-provisioner-${NODE_NAME}-${NODE_ID}`, it is annotated
+in PVs with `AnnProvisionedBy`, to indicate which PVs are up to a certain node.
+To distinguish it from the new env `PROVISIONER_NAME`, we will rename it to `provisionerTag`.
+
+New field will be added into the configuration ConfigMap, to specify available
+volume group for lvm:
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-static:
+      hostDir: "/mnt/ssds"
+      mountDir: "/local-ssds"
+    local-dynamic:
+      lvm:
+        volumeGroup: "volumeGroup1"

```

Comment:
Who creates the volume group? Is it something that admin must do during installation?

Reply:
yeah, sorry the discussion got flooded
> IMHO it's problematic if the provisioner is responsible for this. As mentioned below, disk naming etc could be problematic. Instead, I believe storage pool provisioning is a responsibility of the cluster operator (using e.g. Ansible or whatever), only volume provisioning should be performed by the Kubernetes infrastructure.

> As such, in the LVM case: the Kubernetes provider creates and manages LVs in a given and pre-existing VG, an admin creates PVs and VGs using whichever mechanism seems suitable.

> Agree, there are too many ways an admin can configure and manage vgs. It will be hard to support all of them.

> This will also simplify the issue where all the disks on every node must be named the same.

@msau42 and @NicolasT agree it reasonable to pass in existing groups as input.

Reply:
I am fine with that, however please note that explicitly in the proposal.

Reply:
Can you add a section about the LVM details, and cover some things like:
* the provisioner arguments
* any hostpath volumes lvm requires
* what it will do for each of the backend interfaces
* format of the device name when lv is created
* how to handle lv naming with multiple vgs (multiple storage classes)
* how reboot/restart is handled

Reply:
done